### ISSUE TITLE
Investigate use of Hypothesis for testing

### ISSUE THREAD

**Issue by matthewfeickert** [diana-hep/pyhf#511]
# Description

Given discussions with @Zac-HD at SciPy 2019, it would be interesting to investigate using [Hypothesis](https://github.com/HypothesisWorks/hypothesis) with the pytest test suite.

As @lukasheinrich has pointed out, a good first example to try out would be ensuring that all CLs values are [0,1].

**Comment by kratsg** [comment]
> As @lukasheinrich has pointed out, a good first example to try out would be ensuring that all CLs values are [0,1].

how so?

**Comment by matthewfeickert** [comment]
> how so?

In terms of technical implementation? Or in terms of why it might be useful?

**Comment by kratsg** [comment]
I guess the technical implementation just means letting it generate random workspaces and just ensure the CLs values are always btwn 0 and 1...

**Comment by lukasheinrich** [comment]
I meant it in the way that this is definitely a guarante that we want to provide and hypothesis is all about testing assumed guarantees.

**Comment by lukasheinrich** [comment]
but I think the interpolation testing is a better first example

**Comment by matthewfeickert** [comment]
I haven't fully explored Hypothesis yet (as I only learned about it a week ago). I'm willing to take up poking around and then adding a PR after I finish my defense next week, as the SciPy talk made it seems like it should be easy to get going fast.

**Comment by Zac-HD** [comment]
Hey all!  Thanks for pinging me - I'd also be happy to help review any PRs that you open using Hypothesis.  If you're wondering where to start, I do usually start with "given any valid input, it should never raise an exception" - checking that probabilities are in [0, 1] is really just a cherry on top.  My [tutorial exercises](https://github.com/Zac-HD/escape-from-automanual-testing) might also be helpful if you're wondering where to start or what the tests might look like.

Unfortunately `hypothesis-jsonschema` doesn't (yet) support references; it's on my todo list but hard to do cleanly.

**Comment by lukasheinrich** [comment]
this could work as a simple start for using hypothesis

thoughts @matthewfeickert @kratsg 

```
from hypothesis import given, settings
from hypothesis.strategies import decimals
import pyhf

@given(decimals(1,10),decimals(1,10),decimals(1,10))
@settings(max_examples=5)
def test_cls(s,b,db):
    m = pyhf.simplemodels.hepdata_like([s],[b],[db])
    cls = pyhf.infer.hypotest(1.0,[b]+m.config.auxdata,m)
    assert 0.0 < cls[0] < 1.0
```

**Comment by Zac-HD** [comment]
@lukasheinrich - looks like a great place to start!   A nice extension to that test might be to check that if you run the same computation with different backends, the results are reasonably close.  You'd definitely want more than five examples for that, though personally I'd still with the default even for this :wink: 

And @matthewfeickert `hypothesis-jsonschema` *does* support references now, so that might be a nice way to generate workspaces if you still have schemas for them :smile:

**Comment by lukasheinrich** [comment]
@Zac-HD yeah but we probably need to balance test time against number of examples (this is ~20ms per example.. so maybe 100 examples could work

**Comment by Zac-HD** [comment]
Try running `pytest --hypothesis-show-statistics`, it'll give you a nice timing breakdown including the split between generating data and executing your test code.
